### PR TITLE
feat: nonce in signed messages

### PR DESCRIPTION
**Pull request by mattstam**
RPC messages now accept a signed nonce to verify them against the stored nonce. The goal of signing and verifying nonces is to prevent replay attacks and out-of-order messages.

This is achieved by first calling `get_nonce()` before signing the message, then signing the message that includes it.

### CODE REVIEW

File: sdk/src/client.rs [hunk 1]
```diff
@@ -70,6 +70,17 @@ impl NetworkClient {
...
+    pub async fn get_nonce(&self) -> u64 {
+        let res = self
+            .rpc
+            .get_nonce(GetNonceRequest {
+                address: self.auth.get_address().to_vec(),
+            })
+            .await
+            .unwrap();
+        res.nonce
+    }
```
Comment (ratankaliani) [inline sdk/src/client.rs:82]: docs: Add a more clear specification for what a nonce is and where you're querying it from? i.e. querying from sequencer

File: sdk/src/auth.rs [hunk 2]
```diff
@@ -79,6 +87,7 @@ impl NetworkAuth {
     /// Signs a message to relay a proof.
     pub async fn sign_relay_proof_message(
         &self,
```
Comment (ratankaliani) [inline sdk/src/auth.rs:89]: docs: This is unclear. This should specify that a user is requesting for a hosted relay of a proof.

File: sdk/src/auth.rs [hunk 3]
```diff
@@ -63,10 +70,11 @@ impl NetworkAuth {
     }
 
     /// Signs a message to submit a proof.
```
Comment (ratankaliani) [inline sdk/src/auth.rs:72]: docs: Signs a message to mark a proof as ready for proof generation.

File: sdk/src/auth.rs [hunk 4]
```diff
@@ -50,11 +53,15 @@ impl NetworkAuth {
         }
     }
 
+    pub fn get_address(&self) -> [u8; 20] {
+        *self.wallet.address().0
+    }
+
     /// Signs a message to create a proof.
```
Comment (ratankaliani) [inline sdk/src/auth.rs:60]: docs: Clarify what this fn does.

**Review by ratankaliani — APPROVED** [review]
LGTM, left some small comments on adding more docs for each stage of requesting proof generation which will be helpful as the protocol evolves.

File: sdk/src/client.rs (hunk 1 above)
Comment (mattstam) [reply sdk/src/client.rs:82]: added

File: sdk/src/auth.rs (hunk 3 above)
Comment (mattstam) [reply sdk/src/auth.rs:72]: added

File: sdk/src/auth.rs (hunk 2 above)
Comment (mattstam) [reply sdk/src/auth.rs:89]: added

**Pull request merged**